Mathlib/Tactic/Autolabels/EnvExtension.lean

```diff
@@ -0,0 +1,248 @@
+`add_label label_name dirs: d₁ d₂ ... dₙ`.  This adds
+* `t-label-name` as a label and
+* `Mathlib.d₁, Mathlib.d₂, ..., Mathlib.dₙ` as root folders.
+
+Finally, if you also want to filter out of the previous root folders some subfolders, you can use
+`add_label label_name dirs: d₁ d₂ ... dₙ exclusions: e₁ e₂ ... eₘ`.  This adds
+* `t-label-name` as a label and
+* `Mathlib.d₁, Mathlib.d₂, ..., Mathlib.dₙ` as root folders, unless the folder matches
+* `Mathlib.e₁, Mathlib.e₂, ..., Mathlib.eₘ`.
+
+The helper command `check_labels` displays all the labels that have already been declared.
+-/
+syntax (name := addLabelStx)
+  "add_label " ident ("dirs: " ident*)? ("exclusions: " ident*)? : command
+
+open Elab.Command in
+@[inherit_doc addLabelStx]
+elab_rules : command
+  | `(command| add_label $id dirs: $dirs* exclusions: $excs*) => do
+    setEnv <| labelsExt.addEntry (← getEnv)
+      { label := "t-" ++ id.getId.toString.replace "_" "-"
+        dirs := formatIdsToDirs dirs
+        exclusions := formatIdsToDirs excs }
+  | `(command| add_label $id dirs: $dirs*) => do
+    elabCommand (← `(command| add_label $id dirs: $dirs* exclusions:))
+  | `(command| add_label $id) => do
+    elabCommand (← `(command| add_label $id dirs: $id))
+
+/--
+`check_labels` is a helper command to `add_labels`.
+It displays all the labels that have already been declared.
+`check_labels "t-abc"` shows only the labels starting with `t-abc`.
+-/
+elab "check_labels" st:(ppSpace str)? : command => do
+  let str := (st.getD default).getString
+  for l in labelsExt.getState (← getEnv) do
+    if str.isPrefixOf l.label then
+      logInfo m!"label: {l.label}\ndirs: {l.dirs}\nexclusions: {l.exclusions}"
+
+/--
+`labelsToFiles env gitDiffs` takes as input an `Environment` `env` and a string `gitDiffs`.
+It assumes that `gitDiffs` is a line-break-separated list of paths to files.
+It returns the pairings `Label.label → Array filenames` as a `HashMap`.
+-/
+def labelsToFiles (env : Environment) (gitDiffs : String) : HashMap String (Array String) :=
+  let labels := labelsExt.getState env
+  let gitDiffs := (gitDiffs.splitOn "\n").toArray
+  let hash := addAllLabels gitDiffs labels
+  let grouped := gitDiffs.groupByKey (hash.find? · |>.getD "")
+  let matched := grouped.erase ""
+  let unmatched := ((grouped.find? "").getD #[]).filter (!·.isEmpty)
+  if unmatched.isEmpty then
+      matched
+    else
+      matched.insert "unlabeled" unmatched
+
+/--
+`produceLabels env gitDiffs` takes as input an `Environment` `env` and a string `gitDiffs`.
+It assumes that `gitDiffs` is a line-break-separate list of paths to files.
+It uses the paths to check if any of the labels in the environment is applicable.
+It returns the sorted array of the applicable labels with no repetitions.
+-/
+def produceLabels (env : Environment) (gitDiffs : String) : Array String :=
+  let grouped := labelsToFiles env gitDiffs
+  ((grouped.toArray.map Prod.fst).filter (· != "")).qsort (· < ·)
```

Comment:
Is this sorting alphabetically?

Reply:
Yes, I think so: the keys in `grouped` are `String`s, so `Prod.fst` returns an array of `String`s and then `qsort` uses `<` on `String`s which is alphabetical.